arangod/RestServer/DatabaseFeature.cpp

```diff
@@ -291,6 +294,165 @@ void DatabaseManagerThread::run() {
+    : ServerThread<ArangodServer>(server, "IOHeartbeat"),
+      _exeTimeHistogram(metricsFeature.add(arangodb_ioheartbeat_duration{})),
+      _writeFailures(
+          metricsFeature.add(arangodb_ioheartbeat_failures_write_total{})),
+      _readFailures(
+          metricsFeature.add(arangodb_ioheartbeat_failures_read_total{})),
+      _removeFailures(
+          metricsFeature.add(arangodb_ioheartbeat_failures_remove_total{})),
+      _writeDelays(
+          metricsFeature.add(arangodb_ioheartbeat_delays_write_total{})),
+      _readDelays(metricsFeature.add(arangodb_ioheartbeat_delays_read_total{})),
+      _removeDelays(
+          metricsFeature.add(arangodb_ioheartbeat_delays_remove_total{})) {}
+
+IOHeartbeatThread::~IOHeartbeatThread() { shutdown(); }
+
+void IOHeartbeatThread::run() {
+  auto& databasePathFeature = server().getFeature<DatabasePathFeature>();
+  std::string testFilePath = FileUtils::buildFilename(
+      databasePathFeature.directory().c_str(), "TestBalloonFileIOHeartbeat");
+  std::string testFileContent = "This is just an I/O test.\n";
+
+  LOG_TOPIC("66665", INFO, Logger::ENGINES)
+      << "IOHeartbeatThread: running with checkInterval= " << checkInterval()
+      << " ...";
+
+  unsigned int counter = checkInterval();
+
+  while (true) {
+    try {  // protect thread against any exceptions
+      if (isStopping()) {
+        // done
+        break;
+      }
+
+      if (--counter == 0) {
+        LOG_TOPIC("66659", DEBUG, Logger::ENGINES)
+            << "IOHeartbeat: testing to write/read/remove " << testFilePath;
+        counter = checkInterval();
+        // We simply write a file and sync it to disk in the database
+        // directory and then read it and then delete it again:
+        auto start1 = std::chrono::steady_clock::now();
+        bool trouble = false;
+        try {
+          FileUtils::spit(testFilePath, testFileContent, true);
+        } catch (std::exception const& exc) {
+          ++_writeFailures;
+          LOG_TOPIC("66663", INFO, Logger::ENGINES)
+              << "IOHeartbeat: exception when writing test file: "
+              << exc.what();
+          trouble = true;
+        }
+        auto finish = std::chrono::steady_clock::now();
+        std::chrono::duration<double> dur = finish - start1;
+        bool delayed = dur > std::chrono::seconds(1);
+        if (trouble || delayed) {
+          if (delayed) {
+            ++_writeDelays;
+          }
+          LOG_TOPIC("66662", INFO, Logger::ENGINES)
+              << "IOHeartbeat: trying to write test file took "
+              << std::chrono::duration_cast<std::chrono::microseconds>(dur)
+                     .count()
+              << " milliseconds.";
+        }
+
+        // Read the file if we can reasonably assume it is there:
+        if (!trouble) {
+          auto start = std::chrono::steady_clock::now();
+          try {
+            std::string content = FileUtils::slurp(testFilePath);
```

Comment:
As `content` is not used later...
```suggestion
            FileUtils::slurp(testFilePath);
```

Reply:
I have added a check for the right content. This was forgotten.